### ISSUE TITLE
Convert doesn't cover original face

### ISSUE THREAD

**Issue by Kirin-kun** [iperov/OpenDeepFaceSwap#16]
Original:
![gamer0008b](https://user-images.githubusercontent.com/36699120/39100041-0e4d5e46-4684-11e8-9c0d-434db3b71dcf.png)

converted:
![gamer0008](https://user-images.githubusercontent.com/36699120/39100013-b20db11c-4683-11e8-83e7-898098481845.png)

I trained it until 0.020 with IAEF128 and it looks like this at conversion. 

I'll pass on the blurry ugly face, I'll put that on the training or the model, but it's not covering the face correctly. It's too small and warped weirdly.

Actually, it only looks somewhat correct if the original is looking at the camera and the head is not tilted.

Command line used:

python main.py convert --input-dir H:\Fakes\game --output-dir H:\Fakes\game\merged  --aligned-dir H:\fakes\game\aligned --model-dir h:\fakes\model.game --model IAEF128

**Comment by Kirin-kun** [comment]
It's not related to tilted head actually. It's related to the mask erosion?

With debug:

![head1](https://user-images.githubusercontent.com/36699120/39100230-2d6314f8-4687-11e8-9fd9-dc83ee315a6e.png)
![mask1](https://user-images.githubusercontent.com/36699120/39100232-34f4d12a-4687-11e8-9f54-4d2539c92d73.png)
![mask2](https://user-images.githubusercontent.com/36699120/39100233-37c70f44-4687-11e8-9050-b592f8c3a964.png)

**Comment by HelpSeeker** [comment]
Perhaps check the facial landmarks of the extracted faces. Were all the faces detected correctly?

**Comment by Kirin-kun** [comment]
Everything was detected correctly. It's a face really easy to detect.

**Comment by iperov** [comment]
debug mask doesnt match your result in head post

**Comment by iperov** [comment]
can you upload your workspace ?

**Comment by Kirin-kun** [comment]
>debug mask doesnt match your result in head post

I didn't take the head post's image for debug, because debug doesn't show the result. It only shows original face -> mask -> predicted face -> blurred mask.

For second post it's the same as first post: the eyelashes are showing on the side.

I deleted the model files 3 days ago. I'll redo a training session with same data so I can post more results.

**Comment by iperov** [comment]
just upload your src and dst faces, I will train it for myself

**Comment by Kirin-kun** [comment]
Face forward conversion (new dataset, DF Model):

Before conversion:
![games00219o](https://user-images.githubusercontent.com/36699120/39395734-962e0b3c-4ae2-11e8-8cb2-b63794cd9972.png)
After conversion:
![games00219](https://user-images.githubusercontent.com/36699120/39395738-a44fb314-4ae2-11e8-9ba5-6afd279ffbfe.png)

There's a shadow over the eye (where does the black come from?) and the color of the face is more like the color of the src than the dst.

Side face conversion:
Before:
![games05417o](https://user-images.githubusercontent.com/36699120/39395767-22fa94d6-4ae3-11e8-956e-21484adf8508.png)
After:
![games05417](https://user-images.githubusercontent.com/36699120/39395699-e7872a3c-4ae1-11e8-981e-f4b2fea69142.png)

It doesn't cover the face. It's shifted.

It was all default options conversion.

**Comment by Kirin-kun** [comment]
I could upload the workspace, but it's really large (20/30Gb).

I can give you the links to the videos though. They are both public (though one of them is NSFW).

**Comment by Kirin-kun** [comment]
Or maybe the dark color over the eyes is a learning artifact. It looks likes strand of hair?

![games07030](https://user-images.githubusercontent.com/36699120/39395862-5c332c6c-4ae4-11e8-85bb-05304a1ae0d6.png)

**Comment by Kirin-kun** [comment]
But other than that, it looks nice with DF model.

IAEF128's converted aces look like a mix of both faces, so it's rather unusable.

**Comment by Kirin-kun** [comment]
Also, for some reason, she's cross-eyes. The right pupil has a tendency to be at the inside corner. Again, probably something learned.

**Comment by Kirin-kun** [comment]
But side face conversion sux.

**Comment by iperov** [comment]
did you try convert hist-match ?
convert seamless not always good.

**Comment by iperov** [comment]
> It doesn't cover the face. It's shifted.
> 
> It was all default options conversion.

looks like mask predicted improperly.

**Comment by iperov** [comment]
I found bug in training data generator

before , mask cutted
![python_2018-04-29_00-48-23](https://user-images.githubusercontent.com/8076202/39400722-0ba995c0-4b47-11e8-8d6c-c3966a37cd12.jpg)

after:
![python_2018-04-29_00-49-25](https://user-images.githubusercontent.com/8076202/39400729-32909c2e-4b47-11e8-8ffb-88c92ca926f9.jpg)

**Comment by iperov** [comment]
I hope it will affect result.

**Comment by Kirin-kun** [comment]
Will do.

**Comment by Kirin-kun** [comment]
Still odd result. With debug, it looks almost good just before it applies blur/erosion/seamless, but then the src face shows on the border:

![5317a](https://user-images.githubusercontent.com/36699120/39601823-9203032e-4f23-11e8-92b4-36177bc1e7b6.png)
![5317b](https://user-images.githubusercontent.com/36699120/39601829-96073a9e-4f23-11e8-891c-f7984ef84738.png)

With normal faces, it's ok, but with side faces, it's a hit and miss. Sometimes, it just looks like a mess. Maybe a problem with the values used for ero/blur.

src video: https://www.youtube.com/watch?v=Ve84sjm5vJg&t=35s
dst video: https://bit.ly/2rluqjV (NSFW). I choose these ones because they look straight at the camera most of the time.

**Comment by iperov** [comment]
did you manually crop mask in photoshop in first screen ?

**Comment by iperov** [comment]
anyway, there is nothing to do anymore. Not all fakes will success. Deepfakes highly depends on src and dst data.

**Comment by Kirin-kun** [comment]
In first screen, it was debug mode of convert that put the mask. It shows several windows with each step of the convert (but doesn't actually write the image so I had to do a printscreen). 

The next step (ero/blur/seamless) ate the left side of the mask for whatever reason, so the eyelash of dst shows.

I trained the same dataset with faceswap original and it doesn't have this problem. I still think there's an issue with your automated convert when the zoomed face isn't exactly facing the camera, but if you think this isn't a problem, let's forget it.
